src/main/java/org/gaul/s3proxy/EncryptedBlobStore.java

```diff
@@ -0,0 +1,774 @@
+
+import org.apache.commons.codec.digest.DigestUtils;
+import org.gaul.s3proxy.crypto.Constants;
+import org.gaul.s3proxy.crypto.Decryption;
+import org.gaul.s3proxy.crypto.Encryption;
+import org.jclouds.blobstore.BlobStore;
+import org.jclouds.blobstore.domain.Blob;
+import org.jclouds.blobstore.domain.BlobAccess;
+import org.jclouds.blobstore.domain.BlobBuilder;
+import org.jclouds.blobstore.domain.BlobMetadata;
+import org.jclouds.blobstore.domain.MultipartPart;
+import org.jclouds.blobstore.domain.MultipartUpload;
+import org.jclouds.blobstore.domain.MutableBlobMetadata;
+import org.jclouds.blobstore.domain.PageSet;
+import org.jclouds.blobstore.domain.StorageMetadata;
+import org.jclouds.blobstore.domain.internal.MutableBlobMetadataImpl;
+import org.jclouds.blobstore.domain.internal.PageSetImpl;
+import org.jclouds.blobstore.options.CopyOptions;
+import org.jclouds.blobstore.options.GetOptions;
+import org.jclouds.blobstore.options.ListContainerOptions;
+import org.jclouds.blobstore.options.PutOptions;
+import org.jclouds.blobstore.util.ForwardingBlobStore;
+import org.jclouds.io.ContentMetadata;
+import org.jclouds.io.MutableContentMetadata;
+import org.jclouds.io.Payload;
+import org.jclouds.io.Payloads;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@SuppressWarnings("UnstableApiUsage")
+public final class EncryptedBlobStore extends ForwardingBlobStore {
+    private final Logger logger =
+        LoggerFactory.getLogger(EncryptedBlobStore.class);
+    private SecretKeySpec secretKey;
+
+    private EncryptedBlobStore(BlobStore blobStore, Properties properties)
+        throws IllegalArgumentException {
+        super(blobStore);
+
+        String password = properties.getProperty(
+            S3ProxyConstants.PROPERTY_ENCRYPTED_BLOBSTORE_PASSWORD);
+        checkArgument(!Strings.isNullOrEmpty(password),
+            "Password for encrypted blobstore is not set");
+
+        String salt = properties.getProperty(
+            S3ProxyConstants.PROPERTY_ENCRYPTED_BLOBSTORE_SALT);
+        checkArgument(!Strings.isNullOrEmpty(salt),
+            "Salt for encrypted blobstore is not set");
+        initStore(password, salt);
+    }
+
+    static BlobStore newEncryptedBlobStore(BlobStore blobStore,
+        Properties properties) throws IOException {
+        return new EncryptedBlobStore(blobStore, properties);
+    }
+
+    private void initStore(String password, String salt)
+        throws IllegalArgumentException {
+        try {
+            SecretKeyFactory factory =
+                SecretKeyFactory.getInstance("PBKDF2WithHmacSHA256");
+            KeySpec spec =
+                new PBEKeySpec(password.toCharArray(), salt.getBytes(), 65536,
+                    128);
+            SecretKey tmp = factory.generateSecret(spec);
+            secretKey = new SecretKeySpec(tmp.getEncoded(), "AES");
+        } catch (Exception e) {
+            throw new IllegalArgumentException(e);
+        }
```

Comment:
Why hardcoding a key derivation from a password instead of simply asking for a key ?

I would expect people able to setup s3proxy also able to generate a good key.

Reply:
@fungiboletus well I tried to keep it simple but I am happy if you could open a PR after it is merged that allows "expert" people to set their holy key while using a password should still be possible.

Reply:
I understand the need to keep it simple. You do allow people to use shitty passwords though.